### DISCUSSION TITLE
Log view around a specific commit

### DISCUSSION THREAD

**Discussion by zimoun** [magit/magit#4650] [Ideas]
Hi,

Reading #2990 `Given a commit, show log of "surrounding" commits`, I am not able to do that with Magit.  I am still using a custom alias:

```
around = "!f() { git log --oneline --reverse ${1}~${2}..master | head -n $(echo 2*$2 | bc); }; f"
```
Then `git around 123abc 5` displays some commits surrounding the commit 123abc.  For sure, there are many corner cases for dealing with the general case, but most of the time, a commit is part of a series, so after bisect which finds the culprit, one usually needs to see the history "around" this commit.

From the manual `magit-log-move-to-revision` could fit the job but usually the commit `isn’t visible in the current log buffer`.

And `l o` then `123abc~5...123abc` almost fits the job; but I show only the commits before and I would like to also see some commits after.  Somehow, it is missing a kind of `Type + to show more history` on the top for more recent and the same on the bottom for more older.

@kyleam provided `magit-log-merged` but it is not what I need, IIUC.

Do I miss something?  Is it possible?

Thanks in advance.
Cheers,
simon

**Comment by kyleam** [comment]
> provided magit-log-merged but it is not what I need, IIUC.

`magit-log-merged` uses [git-when-merged](https://github.com/mhagger/git-when-merged) underneath, which is very much tailored to pulling out topic branches that were merged with a dedicated merge commit. For example, pulling the commit chain of a and c out of

    M
    | \
    |  c
    b  |
    |  a
    |/
    |
    .

So, that part doesn't sound like what you're after.

However, as Jonas mentioned in the issue you linked to, he added some logic to show surrounding commits for the case where a commit is made directly on the main line (b above).

<https://github.com/magit/magit/blob/d1ad778280f651f0ab87ce191784b96b3ebe312a/lisp/magit-log.el#L838-L849>

As far as I can see, that aligns pretty well with what you're asking for (though I may be missing something because, while I use `magit-log-merged` regularly, I rarely use the "directly on main line" functionality). As an example, here's what I see with your alias:

    $ git around 29f5be357 10
    95b43253 Do not set point when invoking context-menu
    38cb9372 Add section-specific context-menus
    f3afc117 Bump dependencies on Melpa
    d134b87a magit-ediff-use-indirect-buffers: Demote to variable
    4ff3db8a magit-diff-hunk-line: Visit for diffs that lack a proper hunk
    9a3b8517 templates: Add note about related file
    b1702991 Update README.md and CONTRIBUTING.md
    4b1b7b4a magit-section: Load format-spec
    3960428d Use cl-mapcan instead of mapcan for Emacs 25
    29f5be35 Backport menu functions for Emacs < 28
    7c8254f6 Function-quote many more functions
    ea5d557f magit-section-maybe-update-visibility-indicator: Remove old kludge
    6f5f72bc magit-make-margin-overlay: Remove old kludge
    0154391f make: Rename EMACSBIN variable to EMACS
    d1ad7782 magit-{cherry-pick,revert}-in-progress-p: Restore sequencer/todo check

And here is what I see with `magit-log-merged` via `l-rm 29f5be357 master`:

    95b43253 Do not set point when invoking context-menu
    38cb9372 Add section-specific context-menus
    f3afc117 refs/pull/origin/4645 Bump dependencies on Melpa
    d134b87a magit-ediff-use-indirect-buffers: Demote to variable
    4ff3db8a magit-diff-hunk-line: Visit for diffs that lack a proper hunk
    9a3b8517 templates: Add note about related file
    b1702991 Update README.md and CONTRIBUTING.md
    4b1b7b4a magit-section: Load format-spec
    3960428d Use cl-mapcan instead of mapcan for Emacs 25
    29f5be35 Backport menu functions for Emacs < 28
    7c8254f6 Function-quote many more functions
    ea5d557f magit-section-maybe-update-visibility-indicator: Remove old kludge
    6f5f72bc magit-make-margin-overlay: Remove old kludge
    0154391f make: Rename EMACSBIN variable to EMACS
    d1ad7782 origin/master magit-{cherry-pick,revert}-in-progress-p: Restore sequencer/todo check

**Reply by zimoun** [reply]
Hi,

Thanks!  Indeed, `magit-log-merged` does what I want.  I have tow questions:

 1. What do I configure transient that `m` appears?
 2. What is the way for picking multiple hashes?  As one could do with a rectangular kill/yank.

Cheers,
simon

**Reply by kyleam** [reply]
> What do I configure transient that `m` appears?

`C-x l`.  See https://magit.vc/manual/transient/Enabling-and-Disabling-Suffixes.html#Enabling-and-Disabling-Suffixes

> What is the way for picking multiple hashes? As one could do with a rectangular kill/yank.

Sorry, I'm not following this one.

**Reply by zimoun** [reply]
Thank you.

> `C x l`

Is it possible to share the configuration across various machines ? Via the files `levels` and `values` in the folder `transient/` ?

> >    What is the way for picking multiple hashes? As one could do with a rectangular kill/yank.
>
> Sorry, I'm not following this one.

Back to your examples,

```
95b43253 Do not set point when invoking context-menu
38cb9372 Add section-specific context-menus
f3afc117 refs/pull/origin/4645 Bump dependencies on Melpa
d134b87a magit-ediff-use-indirect-buffers: Demote to variable
4ff3db8a magit-diff-hunk-line: Visit for diffs that lack a proper hunk
9a3b8517 templates: Add note about related file
b1702991 Update README.md and CONTRIBUTING.md
4b1b7b4a magit-section: Load format-spec
3960428d Use cl-mapcan instead of mapcan for Emacs
```
And I would to copy only,

```
d134b87a
4ff3db8a
9a3b8517
```
or their long version.

**Reply by kyleam** [reply]
> Is it possible to share the configuration across various machines ? Via the files `levels` and `values` in the folder `transient/` ?

I haven't done that myself (I configure `transient-default-level` to 7), but, yes, I think that would work.  (See <https://github.com/magit/transient/issues/31>.)

> And I would to copy only,
>
> ```
> d134b87a
> 4ff3db8a
> 9a3b8517
> ```
> or their long version.

I'm not aware of a built-in way to do this.  You could add a custom command to your config, something like:

```elisp
(defun my/magit-copy-region-commits (&optional read-separator)
  (interactive "P")
  (when-let ((commits (magit-region-values 'commit)))
    (deactivate-mark)
    (kill-new
     (message
      "%s"
      (mapconcat #'identity commits
                 (if read-separator (read-string "Separator: ") "\n"))))))
```

**Reply by zimoun** [reply]
Thanks for the both pointers.

Cheers,
simon
